Sources/Action.swift

```diff
@@ -80,39 +88,78 @@ public final class Action<Input, Output, Error: Swift.Error> {
 	///
 	/// - parameters:
 	///   - state: A property to be the state of the `Action`.
-	///   - enabledIf: A predicate which determines the availability of the `Action`,
+	///   - isEnabled: A predicate which determines the availability of the `Action`,
 	///                given the latest `Action` state.
 	///   - execute: A closure that produces a unit of work, as `SignalProducer`, to be
 	///              executed by the `Action`.
-	public init<State: PropertyProtocol>(state property: State, enabledIf isEnabled: @escaping (State.Value) -> Bool, execute: @escaping (State.Value, Input) -> SignalProducer<Output, Error>) {
+	public init<State: PropertyProtocol>(state: State, enabledIf isEnabled: @escaping (State.Value) -> Bool, execute: @escaping (State.Value, Input) -> SignalProducer<Output, Error>) {
+		let isUserEnabled = isEnabled
+
 		deinitToken = Lifetime.Token()
 		lifetime = Lifetime(deinitToken)
-		
-		// Retain the `property` for the created `Action`.
-		lifetime.observeEnded { _ = property }
 
-		executeClosure = { state, input in execute(state as! State.Value, input) }
+		// `isExecuting` has its own `Signal`, so that legitimate feedbacks would
+		// not deadlock.
+		let actionState = MutableProperty(ActionState<State.Value>(isUserEnabled: true, isExecuting: false, value: state.value))
+		self.isEnabled = actionState.map { $0.isEnabled }
+
+		let isExecuting = MutableProperty(false)
+		self.isExecuting = Property(capturing: isExecuting)
```

Comment:
The point of giving it its own backing is that `isExecuting` calls out to the observers outside the signal of `actionState`.

However, it does seem violating exclusivity of access so it might need a change anyway. We might need to add guards in `MutableProperty` to prevent nested modifications too.